audformat/core/utils.py

```diff
@@ -439,3 +441,87 @@ def to_segmented_index(
     obj = obj.reset_index(drop=True)
     obj.index = index
     return obj
+
+
+def union(
+    objs: typing.Sequence[typing.Union[pd.Index]],
+) -> pd.Index:
+    r"""Create union of index objects.
+
+    Index objects must be conform to
+    :ref:`table specifications <data-tables:Tables>`.
+
+    If at least one object is segmented, the output is a segmented index.
+
+    Args:
+        objs: index objects conform to
+            :ref:`table specifications <data-tables:Tables>`
+
+    Returns:
+        union of index objects
+
+    Raises:
+        ValueError: if one or more objects are not conform to
+            :ref:`table specifications <data-tables:Tables>`
+
+    Example:
+        >>> i1 = filewise_index(['f1', 'f2', 'f3'])
+        >>> i2 = filewise_index(['f2', 'f3', 'f4'])
+        >>> union([i1, i2])
+        Index(['f1', 'f2', 'f3', 'f4'], dtype='object', name='file')
+        >>> i3 = segmented_index(
+        ...     ['f1', 'f2', 'f3', 'f4'],
+        ...     [0, 0, 0, 0],
+        ...     [1, 1, 1, 1],
+        ... )
+        >>> i4 = segmented_index(
+        ...     ['f1', 'f2', 'f3'],
+        ...     [0, 0, 1],
+        ...     [1, 1, 2],
+        ... )
+        >>> union([i3, i4])
+        MultiIndex([('f1', '0 days 00:00:00', '0 days 00:00:01'),
+                    ('f2', '0 days 00:00:00', '0 days 00:00:01'),
+                    ('f3', '0 days 00:00:00', '0 days 00:00:01'),
+                    ('f3', '0 days 00:00:01', '0 days 00:00:02'),
+                    ('f4', '0 days 00:00:00', '0 days 00:00:01')],
+                   names=['file', 'start', 'end'])
+
+        >>> union([i1, i2, i3, i4])
+        MultiIndex([('f1', '0 days 00:00:00',               NaT),
+                    ('f1', '0 days 00:00:00', '0 days 00:00:01'),
+                    ('f2', '0 days 00:00:00',               NaT),
+                    ('f2', '0 days 00:00:00', '0 days 00:00:01'),
+                    ('f3', '0 days 00:00:00',               NaT),
+                    ('f3', '0 days 00:00:00', '0 days 00:00:01'),
+                    ('f3', '0 days 00:00:01', '0 days 00:00:02'),
+                    ('f4', '0 days 00:00:00',               NaT),
+                    ('f4', '0 days 00:00:00', '0 days 00:00:01')],
+                   names=['file', 'start', 'end'])
+
+
+    """
+    if not objs:
+        return filewise_index()
+
+    types = [index_type(obj) for obj in objs]
+
+    if len(set(types)) != 1:
+        objs = [to_segmented_index(obj) for obj in objs]
+
+    index = objs[0]
+    for obj in objs[1:]:
+        index = index.union(obj)
+
+    if index_type(index) == define.IndexType.SEGMENTED:
+        # asserts that start and end are of type 'timedelta64[ns]'
+        if index.empty:
+            index = segmented_index()
+        elif index.levels[2].empty:
+            index.set_levels(
+                [pd.to_timedelta([])],
+                level=[2],
+                inplace=True,
+            )
```

Comment:
When can it happen that `index.levels[2].empty`?

Does it also mean if I pass two index objects with missing second level the returned index will then contain a second level?

Reply:
> When can it happen that index.levels[2].empty?

Yeah, took me a while to understand why we need this. The reason we have to add this is that `pandas` stores an empty array if all `end` values are set to `NaT`. And unfortunately `pd.Index.union()` converts from `timedelta` to `datetime` in that case. Don't ask me why they do it, but this fixes it.

> Does it also mean if I pass two index objects with missing second level the returned index will then contain a second level?

As I just said, the second level is not really missing, it's just containing only `pd.NaT` and in that case `pandas` stores and empty array. Probably for efficiency, but not sure.

Reply:
OK, makes all sense, but please add a comment explaining why we do it to the code.

Reply:
Honestly, the way pandas handles `timedelta` and `datetime` does not make much sense to me :)

I now added a comment explaining why we need this.

Reply:
Yes, I also don't like how `pandas` handles it.